plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1270 @@
+const BETA: f64 = 0.85;
+
+// This is the `K` from section "5.3 Arrival-time filter" in expression:
+// "where `f_max = max {1/(inter_arrival_time[T(j) - T(j-1))]}` for j in i-K+1,...,i"
+const LAST_INTER_DEPARTURE_SIZE: usize = 25;
+
+// From "5.5 Rate control"
+// It is RECOMMENDED to measure this average and standard deviation with an
+// exponential moving average with the smoothing factor 0.95, as it is
+// expected that this average covers multiple occasions at which we are
+// in the Decrease state.
+const MOVING_AVERAGE_SMOOTHING_FACTOR: f64 = 0.5; // FIXME: Should be 0.95
+
+// `N(i)` is the number of packets received the past T seconds and `L(j)` is
+// the payload size of packet j.  A window between 0.5 and 1 second is
+// RECOMMENDED.
+static PACKETS_RECEIVED_WINDOW: Lazy<Duration> = Lazy::new(|| Duration::milliseconds(1000)); // ms
+
+// from "5.4 Over-use detector" ->
+// Moreover, del_var_th(i) SHOULD NOT be updated if this condition
+// holds:
+//
+// ```
+// |m(i)| - del_var_th(i) > 15
+// ```
+static MAX_M_MINUS_DEL_VAR_TH: Lazy<Duration> = Lazy::new(|| Duration::milliseconds(15));
+
+// from 5.4 "It is also RECOMMENDED to clamp del_var_th(i) to the range [6, 600]"
+static MIN_THRESHOLD: Lazy<Duration> = Lazy::new(|| Duration::milliseconds(6));
+static MAX_THRESHOLD: Lazy<Duration> = Lazy::new(|| Duration::milliseconds(600));
+
+// From 5.5 ""Close" is defined as three standard deviations around this average"
+const STANDARD_DEVIATION_CLOSE_NUM: f64 = 3.;
+
+// Minimal duration between 2 updates on the lost based rate controller
+static LOSS_UPDATE_INTERVAL: Lazy<time::Duration> = Lazy::new(|| time::Duration::from_millis(200));
+static LOSS_DECREASE_THRESHOLD: f64 = 0.1;
+static LOSS_INCREASE_THRESHOLD: f64 = 0.02;
+static LOSS_INCREASE_FACTOR: f64 = 1.05;
+
+// Minimal duration between 2 updates on the lost based rate controller
+static DELAY_UPDATE_INTERVAL: Lazy<time::Duration> = Lazy::new(|| time::Duration::from_millis(100));
+
+static ROUND_TRIP_TIME_WINDOW_SIZE: usize = 100;
+
+fn ts2dur(t: gst::ClockTime) -> Duration {
+    Duration::nanoseconds(t.nseconds() as i64)
+}
+
+fn dur2ts(t: Duration) -> gst::ClockTime {
+    gst::ClockTime::from_nseconds(t.num_nanoseconds().unwrap() as u64)
+}
+
+#[derive(Debug)]
+enum BandwidthEstimationOp {
+    /// Don't update target bitrate
+    Hold,
+    /// Decrease target bitrate
+    Decrease {
+        #[allow(dead_code)]
+        reason: String, // for Debug
+    },
+    Increase,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum ControllerType {
+    // Running the "delay-based controller"
+    Delay,
+    // Running the "loss based controller"
+    Loss,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Packet {
+    departure: Duration,
+    arrival: Duration,
+    size: usize,
+}
+
+fn human_kbits<T: Into<f64>>(bits: T) -> String {
+    format!("{:.2}kb", (bits.into() / 1_000.))
+}
+
+impl Packet {
+    fn from_structure(structure: &gst::Structure) -> Option<Self> {
```

Comment:
```suggestion
    fn from_structure(structure: &gst::StructureRef) -> Option<Self> {
```

Maybe avoid creating a full structure copy where it's used?

Reply:
I need to read more about the Structure vs StructureRef (and friends) difference, it is still not clear in my head it seems :-)

Reply:
`Structure` is an owned structure (you're going to free it on `Drop`), `&StructureRef` is a borrowed structure. It's like the difference between `String` and `&str`, `PathBuf` and `&Path`, etc.

For example, if you have a `glib::Value` and get a `Structure` out of it then it's your own copy of that, if you get a `&StructureRef` out of it then you're only borrowing the one owned by the value.

Reply:
Makes perfect sense, thanks :-)